src/uast.cc

```diff
@@ -24,27 +28,41 @@ struct Uast {
   NodeIface iface;
 };
 
+struct UastIterator {
+  const Uast *ctx;
+  TreeOrder order;
+  std::deque<void *> dqueue;
+  std::set<void *> childrenAdded;
+};
+
 struct Nodes {
-  void **results;
+  std::vector<void *> results;
   int len;
   int cap;
 };
 
 static xmlDocPtr CreateDocument(const Uast *ctx, void *node);
 static xmlNodePtr CreateXmlNode(const Uast *ctx, void *node, xmlNodePtr parent);
 void Error(void *ctx, const char *msg, ...);
+// Adds the children of the node to the iterator queue and returns
+// if the node was already checked, which will happen with leaf nodes
+// or nodes which childs already processed. Used for the POSTORDER
+// iterative traversal algorithm.
+static bool MaybeAddChildren(UastIterator *iter, void *node);
```

Comment:
Yes, wasn't very happy about the name, but forgot to rename it later. The function needs to check if the children were already added and add them if not. Only nodes without children or with already added children (that thus has been returned before) will be returned by `Next` (because if not the children will be before in the double queue and thus will need to be processed before), so the check is actually the important part. Maybe `Visited` is the best name?

Reply:
I don't think the name is too bad, it's just that I find it weird that it returns true when it hasn't added any children. Returning true in that case would make more sense.